compiler/optimizations/deadCodeElimination.cpp

```diff
@@ -490,13 +490,97 @@ static bool isIndexVariable(Symbol* sym, CForLoop* loop) {
   return false;
 }
 
-static  CallExpr* generateGPUCall(FnSymbol* kernel,
-                                     std::vector<VarSymbol*> actuals) {
+static Symbol* extractUpperBoundFromLoop(CForLoop* loop) {
+  if(BlockStmt* bs = toBlockStmt(loop->testBlockGet())) {
+      if(CallExpr *call = toCallExpr(bs->body.head)) {
+        if(call->isPrimitive(PRIM_LESSOREQUAL)) {
+          if(SymExpr *symExpr = toSymExpr(call->get(2))) {
+            return symExpr->symbol();
+          }
+        }
+      }
+  }
+
+  INT_FATAL("Unexpected upper bound for loop identified for extraction as GPU kernel");
+  return nullptr;
+}
+
+static Symbol* extractLowerBoundFromLoop(CForLoop* loop) {
+  if(BlockStmt* bs = toBlockStmt(loop->initBlockGet())) {
+    for_alist(entry, bs->body) {
+      if(CallExpr *call = toCallExpr(entry)) {
+        if(call->isPrimitive(PRIM_ASSIGN)) {
+          if(SymExpr *symExpr = toSymExpr(call->get(2))) {
+            return symExpr->symbol();
+          }
+        }
+      }
+    }
+  }
+
+  INT_FATAL("Unexpected lower bound for loop identified for extraction as GPU kernel");
+  return nullptr;
+}
+
+static VarSymbol* generateNewVariable(BlockStmt* insertionPoint, const char* name, Type* type) {
+  VarSymbol *var = new VarSymbol(name, type);
+  var->defPoint = new DefExpr(var);
+  insertionPoint->insertAtTail(var->defPoint);
+  return var;
+}
+
+static VarSymbol* generateBlockSizeComputation(BlockStmt* gpuLaunchBlock, CForLoop* loop) {
+  Symbol* ub = extractUpperBoundFromLoop(loop);
+  Symbol* lb = extractLowerBoundFromLoop(loop);
+
+  VarSymbol *varBoundDelta = generateNewVariable(gpuLaunchBlock, "blockDelta", dtInt[INT_SIZE_DEFAULT]);
+  VarSymbol *varBlockSize = generateNewVariable(gpuLaunchBlock, "blockSize", dtInt[INT_SIZE_DEFAULT]);
+
+  CallExpr *c1 = new CallExpr(PRIM_ASSIGN, varBoundDelta, new CallExpr(PRIM_SUBTRACT, ub, lb));
+  gpuLaunchBlock->insertAtTail(c1);
+
+  CallExpr *c2 = new CallExpr(PRIM_ASSIGN, varBlockSize, new CallExpr(PRIM_ADD, varBoundDelta, new_IntSymbol(1)));
+  gpuLaunchBlock->insertAtTail(c2);
+
+  return varBlockSize;
+}
+
+static VarSymbol* generateAssignmentToPrimitive(
+  FnSymbol* fn, const char *varName, PrimitiveTag prim, Type *primReturnType) {
+
+  VarSymbol *var = generateNewVariable(fn->body, varName, primReturnType);
+  CallExpr *c1 = new CallExpr(PRIM_MOVE, var, new CallExpr(prim));
+  fn->insertAtTail(c1);
+  return var;
+}
+
+static VarSymbol* generateIndexComputation(FnSymbol* fn, Symbol* sym) {
```

Comment:
Could you add a header comment here describing the AST you write and where `varBlockIdxX`, `t0` etc fall into in that AST? I think you had a pseudocode when we were screensharing, something simple like that should suffice.

Reply:
I'll add the following comment above the function:

```c++
/**
 *  Generates and inserts the following AST into fn:
 *
 *  blockIdxX  = __primitve('gpu blockIdx x')
 *  blockDimX  = __primitve('gpu blockDim x')
 *  threadIdxX = __primitve('gpu threadIdx x')
 *  t0 = varBlockIdxX * varBlockDimX 
 *  t1 = t0 + threadIdxX
 **/
```